README.md
# http-websocket

[![deno land](http://img.shields.io/badge/available%20on-deno.land/x-lightgrey.svg?logo=deno)](https://deno.land/x/http_websocket)
[![deno doc](https://doc.deno.land/badge.svg)](https://doc.deno.land/https/deno.land/x/http_websocket/mod.ts)
[![GitHub](https://img.shields.io/github/license/httpland/http-websocket)](https://github.com/httpland/http-websocket/blob/main/LICENSE)

HTTP request for websocket with standard `Request` and `Response`.

## HTTP handler for websocket

Create WebSocket request handler.

```ts
import {
  createHandler,
  SocketHandler,
} from "https://deno.land/x/http_websocket@$VERSION/mod.ts";
import { serve } from "https://deno.land/std@$VERSION/http/mod.ts";
const socketHandler: SocketHandler = (socket) => {
  socket.onopen = () => {
    socket.send("hello");
  };
};
const handler = createHandler(socketHandler);
await serve(handler);
```

## Select sub-protocol

You can select the websocket sub-protocol. Given a list of sub-protocols from
the client, you can select any sub-protocol you like or none.

client:

```ts
const ws = new WebSocket("ws://localhost/chat", ["soap", "wamp"]);

ws.onopen = () => {
  if (ws.protocol !== "wamp") {
    ws.close();
  }
};
```

request:

```http
GET /chat
Host: localhost
Upgrade: websocket
Connection: Upgrade
Origin: http://localhost
Sec-WebSocket-Key: <SEC_WEBSOCKET_KEY>
Sec-WebSocket-Version: 13
Sec-WebSocket-Protocol: soap, wamp
```

server:

```ts
import {
  createHandler,
  SocketHandler,
} from "https://deno.land/x/http_websocket@$VERSION/mod.ts";
import { serve } from "https://deno.land/std@$VERSION/http/mod.ts";

const handler = createHandler(() => {}, {
  protocol: (protocols) => {
    // protocols: ["soap","wamp"]
    if (protocols.includes("wamp")) {
      return "wamp";
    }
  },
});
await serve(handler);
```

response:

```http
101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: <SEC_WEBSOCKET_ACCEPT>
Sec-WebSocket-Protocol: wamp
```

## Handler Spec

The `Response` includes status code and headers as follow:

| Code  | Headers                                                                   |
| ----- | ------------------------------------------------------------------------- |
| `101` | `upgrade`, `connection`, `sec-websocket-accept`, `sec-websocket-protocol` |
| `400` |                                                                           |
| `405` | `allow`                                                                   |
| `500` |                                                                           |

## WebSocket Status Code and Status Text

Helper for processing status code and status text.

```ts
import {
  Status,
  STATUS_TEXT,
} from "https://deno.land/x/http_websocket@$VERSION/mod.ts";
import { assertEquals } from "https://deno.land/std@$VERSION/testing/asserts.ts";

assertEquals(Status.NormalClosure, 1000);
assertEquals(STATUS_TEXT[Status.NormalClosure], "Normal Closure");
```

## Validate request

HTTP request validation for websocket. The validation is based on RFC 6455, 4.1.

The result of the validation is returned as a tuple. If there is an error, the
second element is [`HttpError`](https://deno.land/std/http/mod.ts?s=HttpError).

```ts
import {
  validateRequest,
} from "https://deno.land/x/http_websocket@$VERSION/mod.ts";
import {
  assertEquals,
  assertIsError,
} from "https://deno.land/std@$VERSION/testing/asserts.ts";

const req = new Request("http://localhost/");
const result = validateRequest(req);
assertEquals(result[0], false);
assertIsError(result[1]);
```

## License

Copyright © 2022-present [httpland](https://github.com/httpland).

Released under the [MIT](./LICENSE) license


CHANGELOG.md
# [1.0.0-beta.5](https://github.com/httpland/http-websocket/compare/1.0.0-beta.4...1.0.0-beta.5) (2022-08-19)


### Features

* **handler:** improve types and docs comment ([546e3a4](https://github.com/httpland/http-websocket/commit/546e3a4d724cb4f43eac989a0c6be1d3f8486d9d))

# [1.0.0-beta.4](https://github.com/httpland/http-websocket/compare/1.0.0-beta.3...1.0.0-beta.4) (2022-08-10)


### Features

* rename module name and add docs comment ([8b8b17b](https://github.com/httpland/http-websocket/commit/8b8b17b8f720110cb6063be4df5fe09b6a0d2ffb))

# [1.0.0-beta.3](https://github.com/TomokiMiyauci/ws-handler/compare/1.0.0-beta.2...1.0.0-beta.3) (2022-08-08)


### Features

* **handler:** add status text to response ([ba82688](https://github.com/TomokiMiyauci/ws-handler/commit/ba826886082a3f31b9e64401a9eff62fbe609912))

# [1.0.0-beta.2](https://github.com/TomokiMiyauci/ws-handler/compare/1.0.0-beta.1...1.0.0-beta.2) (2022-08-08)


### Features

* **header:** add `protocol` and `debug` field to interface ([a186cde](https://github.com/TomokiMiyauci/ws-handler/commit/a186cde7b4396ba8e80e46a46458e1744aea3b59))

# 1.0.0-beta.1 (2022-08-04)


### Bug Fixes

* add default options ([81afc0d](https://github.com/TomokiMiyauci/ws-handler/commit/81afc0dc5a2d4b30483fc67db1043d042ac1dbf1))


### Features

* add `createHandler` and utilities ([dccbd62](https://github.com/TomokiMiyauci/ws-handler/commit/dccbd62296d54fb6f0687c412bd4248d3209763c))


status.ts
/** Standard WebSocket status codes.
 * RFC 6455, 7.4.1
 */
export enum Status {
  NormalClosure = 1000,
  GoingAway = 1001,
  ProtocolError = 1002,
  UnsupportedData = 1003,
  Reserved = 1004,
  NoStatusRcvd = 1005,
  AbnormalClosure = 1006,
  InvalidFramePayloadData = 1007,
  PolicyViolation = 1008,
  MessageTooBig = 1009,
  MandatoryExt = 1010,
  InternalServerError = 1011,
  TLSHandshake = 1015,
}

/** A record of all the status codes text.
 * ```ts
 * import {
 *   Status,
 *   STATUS_TEXT,
 * } from "https://deno.land/x/http_websocket@$VERSION/mod.ts";
 * import { assertEquals } from "https://deno.land/std@$VERSION/testing/asserts.ts";
 * assertEquals(Status.NormalClosure, 1000);
 * assertEquals(STATUS_TEXT[Status.NormalClosure], "Normal Closure");
 * ```
 */
export const STATUS_TEXT: Readonly<Record<Status, string>> = {
  [Status.NormalClosure]: "Normal Closure",
  [Status.GoingAway]: "Going Away",
  [Status.ProtocolError]: "Protocol error",
  [Status.UnsupportedData]: "Unsupported Data",
  [Status.Reserved]: "---Reserved----",
  [Status.NoStatusRcvd]: "No Status Rcvd",
  [Status.AbnormalClosure]: "Abnormal Closure",
  [Status.InvalidFramePayloadData]: "Invalid frame payload data",
  [Status.PolicyViolation]: "Policy Violation",
  [Status.MessageTooBig]: "Message Too Big",
  [Status.MandatoryExt]: "Mandatory Ext.",
  [Status.InternalServerError]: "Internal Server Error",
  [Status.TLSHandshake]: "TLS handshake",
};


handler.ts
import { Handler, safeSync, Status, STATUS_TEXT, trim } from "./deps.ts";
import { validateRequest } from "./validate.ts";

/** WebSocket handler options. */
export interface HandlerOptions {
  /** Select sub-protocol.  */
  readonly protocol?: (protocols: ReadonlyArray<string>) => string | undefined;

  /** Whether to do as a debug mode.
   * When this is `true`, if an error occurs in the server, the
   * Include detailed error messages in the response.
   * @remarks For production use, this feature is not recommended.
   */
  readonly debug?: boolean;

  /**
   * If the client does not respond to this frame with a
   * `pong` within the timeout specified, the connection is deemed
   * unhealthy and is closed. The `close` and `error` event will be emitted.
   *
   * The default is 120 seconds. Set to 0 to disable timeouts.
   */
  readonly idleTimeout?: number;
}

/** Socket handler. */
export type SocketHandler = (
  socket: WebSocket,
  ctx: SocketHandlerContext,
) => Promise<void> | void;

/** Socket handler context. */
export interface SocketHandlerContext {
  /** HTTP `Request` Object */
  readonly request: Request;

  /** HTTP `Response` Object that upgrade to websocket. */
  readonly response: Response;
}

/** Create WebSocket request handler.
 * ```ts
 * import {
 *   createHandler,
 *   SocketHandler,
 * } from "https://deno.land/x/http_websocket@$VERSION/mod.ts";
 * import { serve } from "https://deno.land/std@$VERSION/http/mod.ts";
 * const socketHandler: SocketHandler = (socket) => {
 *   socket.onopen = () => {
 *     socket.send("hello");
 *   };
 * };
 * const handler = createHandler(socketHandler);
 * await serve(handler);
 * ```
 */
export function createHandler(
  socketHandler: SocketHandler,
  { idleTimeout, protocol: _protocol, debug }: HandlerOptions = {},
): Handler {
  return async (req) => {
    const [valid, error] = validateRequest(req);
    if (!valid) {
      const body = debug ? error.message : null;
      const headers = error.headers;
      return new Response(body, {
        status: error.status,
        statusText: STATUS_TEXT[error.status],
        headers,
      });
    }

    const secWebsocketProtocol = req.headers.get("sec-websocket-protocol");
    const protocols = headerList(secWebsocketProtocol);

    const [data, e] = safeSync(() => {
      const protocol = _protocol?.(protocols);

      return Deno.upgradeWebSocket(req, { protocol, idleTimeout });
    });

    if (!data) {
      const body = debug ? resolveErrorMsg(e) : null;
      const res = new Response(body, ErrorResponseInit);
      return res;
    }
    const { socket, response } = data;

    try {
      await socketHandler(socket, {
        request: req.clone(),
        response: response.clone(),
      });
      return response;
    } catch (error) {
      const body = debug ? resolveErrorMsg(error) : null;
      return new Response(body, ErrorResponseInit);
    }
  };
}

function resolveErrorMsg(
  value: unknown,
  fallbackMessage = "Unknown error has occurred",
): string {
  return value instanceof Error ? value.message : fallbackMessage;
}

const ErrorResponseInit: ResponseInit = {
  status: Status.InternalServerError,
  statusText: STATUS_TEXT[Status.InternalServerError],
};

function headerList(header: string | null): string[] {
  const headerList = header?.split(",") ?? [];

  return headerList.map(trim).filter((value) => !!value);
}


validate.ts
import { createHttpError, HttpError, isNull, Status } from "./deps.ts";

/** HTTP request validation for websocket.
 * The validation is based on RFC 6455, 4.1.
 * ```ts
 * import {
 *   validateRequest,
 * } from "https://deno.land/x/http_websocket@$VERSION/mod.ts";
 * import {
 *   assertEquals,
 *   assertIsError,
 * } from "https://deno.land/std@$VERSION/testing/asserts.ts";
 *
 * const req = new Request("http://localhost/");
 * const result = validateRequest(req);
 * assertEquals(result[0], false);
 * assertIsError(result[1]);
 * ```
 */
export function validateRequest(req: Request): [valid: true] | [
  valid: false,
  error: HttpError,
] {
  if (req.method !== "GET") {
    return [
      false,
      createHttpError(Status.MethodNotAllowed, `Invalid HTTP Request method.`, {
        headers: {
          Allow: "GET",
        },
      }),
    ];
  }

  const upgrade = req.headers.get("upgrade");

  if (isNull(upgrade)) {
    return [
      false,
      createHttpError(
        Status.BadRequest,
        `Missing header. "Upgrade"`,
      ),
    ];
  }

  if (upgrade.toLowerCase() !== "websocket") {
    return [
      false,
      createHttpError(
        Status.BadRequest,
        `Invalid header. "Upgrade" header must be "websocket"`,
      ),
    ];
  }

  const connection = req.headers.get("connection");
  if (isNull(connection)) {
    return [
      false,
      createHttpError(Status.BadRequest, `Missing header. "Connection"`),
    ];
  }
  if (connection.toLowerCase() !== "upgrade") {
    return [
      false,
      createHttpError(
        Status.BadRequest,
        `Invalid header. "Connection" header must be "Upgrade"`,
      ),
    ];
  }
  const secWebSocketKey = req.headers.get("Sec-WebSocket-Key");
  if (isNull(secWebSocketKey)) {
    return [
      false,
      createHttpError(Status.BadRequest, `Missing header. "Sec-WebSocket-Key"`),
    ];
  }

  const secWebSocketVersion = req.headers.get("Sec-WebSocket-Version");
  if (isNull(secWebSocketVersion)) {
    return [
      false,
      createHttpError(
        Status.BadRequest,
        `Missing header. "Sec-WebSocket-Version"`,
      ),
    ];
  }
  if (secWebSocketVersion !== "13") {
    return [
      false,
      createHttpError(
        Status.BadRequest,
        `Invalid header. "Sec-WebSocket-Version" must be "13"`,
      ),
    ];
  }

  return [true];
}
